transaction/core/src/tx_summary.rs

```diff
@@ -0,0 +1,210 @@
+    /// transaction
+    #[prost(message, repeated, tag = "1")]
+    pub outputs: Vec<TxOutSummary>,
+
+    /// Data in the summary associated to each real input
+    #[prost(message, repeated, tag = "2")]
+    pub inputs: Vec<TxInSummary>,
+
+    /// Fee paid to the foundation for this transaction
+    #[prost(uint64, tag = "3")]
+    pub fee: u64,
+
+    /// Token id for the fee output of this transaction
+    #[prost(uint64, tag = "4")]
+    pub fee_token_id: u64,
+
+    /// The block index at which this transaction is no longer valid.
+    #[prost(uint64, tag = "5")]
+    pub tombstone_block: u64,
+}
+
+impl TxSummary {
+    /// Make a TxSummary for a given TxPrefix and pseudo-output commitments
+    pub fn new(
+        tx_prefix: &TxPrefix,
+        pseudo_output_commitments: &[CompressedCommitment],
+    ) -> Result<Self, ZipExactError> {
+        // Scratch which helps us associate outputs to inputs with rules
+        let mut input_rules_associated_target_keys: BTreeSet<CompressedRistrettoPublic> =
+            Default::default();
+
+        // Compute the inputs
+        let inputs: Vec<TxInSummary> =
+            zip_exact(tx_prefix.inputs.iter(), pseudo_output_commitments.iter())?
+                .map(|(input, commitment)| {
+                    let mut result = TxInSummary {
+                        pseudo_output_commitment: commitment.clone(),
+                        ..Default::default()
+                    };
+                    if let Some(rules) = &input.input_rules {
+                        result.has_input_rules = true;
+                        let mut associated_keys = rules.associated_tx_target_keys();
+                        input_rules_associated_target_keys.append(&mut associated_keys);
+                    }
+                    result
+                })
+                .collect();
+
+        let outputs: Vec<TxOutSummary> = tx_prefix
+            .outputs
+            .iter()
+            .map(|src| TxOutSummary {
+                masked_amount: src.masked_amount.clone(),
+                target_key: src.target_key,
+                public_key: src.public_key,
+                associated_to_input_rules: input_rules_associated_target_keys
+                    .contains(&src.target_key),
+            })
+            .collect();
+
+        Ok(Self {
+            outputs,
+            inputs,
+            fee: tx_prefix.fee,
+            fee_token_id: tx_prefix.fee_token_id,
+            tombstone_block: tx_prefix.tombstone_block,
+        })
+    }
+}
+
+/// A subset of the data of a TxOut.
+///
+/// Fog hint and memo are omitted to reduce size and complexity on HW device,
+/// which can't really do much with those and isn't very interested in them
+/// anyways.
+#[derive(Clone, Deserialize, Digestible, Eq, Hash, Message, PartialEq, Serialize, Zeroize)]
+pub struct TxOutSummary {
+    /// The amount being sent.
+    #[prost(oneof = "MaskedAmount", tags = "1, 2")]
```

Comment:
this actually has to be 1,6 because i'm using the Enum from `transaction-core` and these tags are also duplicated in that enum.